### ISSUE TITLE
Naming conventions

### ISSUE THREAD

**Issue by timrijkse** [Blocklevel/blue#6]
Let's use this ticket for discussing naming conventions. Does anyone have an idea about this?

**Comment by MatteoGabriele** [comment]
Little bit too generic as a question. Naming conventions for what?

**Comment by timrijkse** [comment]
In the team meet we agreed upon discuss about conventions first before continuing the web pack config stuff.

**Comment by MatteoGabriele** [comment]
I have some points that we can discuss:

- **commits**: For what concerns commits, the #34 is merged, so we have now a unique way to write commit that we should use 

- **feature branches**: I also think we should start create feature differently. First create a new issue, then the name of the feature branch is simply "feature-{issue_number}", so "feature-23" for example. Even tho we mostly open and close the feature branch, it's better to have a branch that is documented and more "followable". Also numbers are quicker, you don't have to come up with a nice name every time 😄 

- **folders and files**: I think we should name folders and files all lowercase and hyphenated, like @supr3m0 suggested, because it's easier to read and also different operating systems can have problems detecting lowercase to uppercase letters.

**Comment by jonaskuiler** [comment]
I agree with @MatteoGabriele. Branches should always be connected to an issue, this is also how JIRA handles it. 

I also want to start using this -> http://jsonapi.org/ but maybe that's more of a back-end issue.

**Comment by floristenhove** [comment]
Yeah, I think it's something related to both back- and frontend.

I came across [this package](https://github.com/nilportugues/laravel5-jsonapi) for Laravel implementing this spec, but doesn't seem to be well maintained any longer.

I'll have a look if this can be easily integrated.

**Comment by reblim** [comment]
Maybe setting a prefix like '**bl-**' (e.g. **bl-header**) for all components, classes and id's will be very helpful.

**Comment by reblim** [comment]
Maybe setting a prefix like '**bl-**' (e.g. **bl-header**) for all components, classes and id's will be very helpful.

**Comment by MatteoGabriele** [comment]
I personally don't like prefixing that much, to be honest.
We only have to avoid HTML5 names. 
Instead of `header.js` we can call it `main-header.js` or `nav-bar.js`. I find prefixing pretty redundant. 

Classes don't need to have names, they should maintain the `$style.page` or `$style.component` in the root element, because the actual name of the class comes from the file itself. In this way we also know immediately if a component is in the `page` folder and so on.

We don't need to use id's because css-modules generates a class that acts like unique id.

**Comment by reblim** [comment]
Fare enough. We could then simply agree on a [semantic naming convention](https://css-tricks.com/semantic-class-names/) to make each project equally navigable throughout. 

Interesting reads on the same topic:
http://thesassway.com/advanced/modular-css-naming-conventions
https://css-tricks.com/semantic-class-names/
http://thesassway.com/intermediate/avoid-nested-selectors-for-more-modular-css

I fully agree with the main component class name, it should stay as `$style.page` or `$style.component`.  

Yeah! let's try to stay away from id's, and specially when using a system like vue, or similar methodologies like BEM or SMACSS.

**Comment by MatteoGabriele** [comment]
example of a **news** component should be something like this:

news.css file

```css
.component {
  padding: 20px;
  width: 200px;
  min-height: 200px;
}

.title {
  color: red;
}

.subtitle {
  color: green;
}

.body {
  color: black;
}
```

news.vue file

```html
<script src="./news.js"></script>

<style module src="./news.css"></style>

<template>
  <div :class="$style.component">
    <h1 :class="$style.title">Lorem ipsum</h1>
    <h2 :class="$style.subtitle">Lorem ipsum dolor sit.</h2>
    <p :class="$style.body">Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
  </div>
</template>
```

The rendered element will be like this.
I like the fact that I can spot easily where this file is located or what it is.

```html
<div class="news__component__1j6z5">
  <h1 class="news__title__2WboV">Lorem ipsum</h1> 
  <h2 class="news__subtitle__36YM6">Lorem ipsum dolor sit.</h2> 
  <p class="news__body__3SquF">Lorem ipsum dolor sit amet, consectetur adipisicing elit.</p>
</div>
```
